File: backend/src/main/java/com/morris/opensquare/services/NotificationService.java
package com.morris.opensquare.services;

import com.morris.opensquare.models.Notifications.GlobalNotification;
import com.morris.opensquare.models.Notifications.OwaspBlogReference;

import java.util.List;

public interface NotificationService {

    /**
     * Broadcast a global notification message across the Opensquare platform to all users.
     *
     * @param message {@link String} global message
     *
     * @return {@link GlobalNotification}
     */
    GlobalNotification broadcastNotification(String message);

    /**
     * Reads all current global notifications.
     *
     * @return {@link List<GlobalNotification>}
     */
    List<GlobalNotification> readAllGlobalNotifications();

    /**
     * Broadcast a global notification message that contains owasp reference acroos the
     * Opensquare platform to all users.
     *
     * @param message {@link String} global owasp message
     * @param owaspBlogReference {@link OwaspBlogReference}
     *
     * @return {@link GlobalNotification}
     */
    GlobalNotification broadcastOwaspNotification(String message, OwaspBlogReference owaspBlogReference);
}

File: backend/src/main/kotlin/com/morris/opensquare/services/NotificationService.kt
package com.morris.opensquare.services

import com.morris.opensquare.models.Notifications.GlobalNotification
import com.morris.opensquare.models.Notifications.OwaspBlogReference
import com.morris.opensquare.repositories.GlobalNotificationsRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.time.LocalDateTime
import java.time.ZoneOffset

@Service
class NotificationService @Autowired constructor(private val globalNotificationsRepository: GlobalNotificationsRepository) {

    companion object {
        private const val SENDER = "Opensquare Team"
        private const val PLUS_DAYS: Long = 5
    }

    fun broadcastNotification(message: String): GlobalNotification {
        val notification = buildSimpleBroadCastNotification(message)
        return globalNotificationsRepository.insert(notification)
    }

    fun readAllGlobalNotifications(): List<GlobalNotification> {
        return globalNotificationsRepository.findAll()
    }

    fun broadcastOwaspNotification(message: String, owaspReference: OwaspBlogReference): GlobalNotification {
        val notification = buildSimpleBroadCastNotification(message)
        notification.owaspRef = owaspReference
        return globalNotificationsRepository.insert(notification)
    }

    private fun buildSimpleBroadCastNotification(message:String): GlobalNotification {
        val weekExpiration = LocalDateTime.now().plusDays(PLUS_DAYS)
        return GlobalNotification.Builder()
            .expiration(weekExpiration.atOffset(ZoneOffset.UTC).toLocalDateTime())
            .message(message)
            .sender(SENDER)
            .build()
    }
}